Introducing react flow

OsamaMSoliman opened the pull request.

Commit:
Pipeline Rendered. TODO Customization

Commit:
Pipeline Tree Chart Done

Commit:
Pipeline Tree Chart Done

Commit:
Pipeline Graph Done

Commit:
Pipeline Graph show in Model Card Done

Commit:
new structure for pipeline details in model card

Commit:
Pipeline Tree and Graph Done. TODO: compare and select any one to keep

OsamaMSoliman committed:
initial reactflow test

OsamaMSoliman committed:
using reactflow hooks

OsamaMSoliman committed:
added root node to the graph

OsamaMSoliman committed:
extracting AnyKeyValuePairs in a separate file and adding http:// to solve the error in the url

OsamaMSoliman committed:
better handling for the axios http protocal problem

OsamaMSoliman committed:
reactflow mock graph in main app

OsamaMSoliman committed:
new pipeline (following Moinam's idea)

OsamaMSoliman committed:
all nodes are interactive and shows config
graphs animate zooming in and out

OsamaMSoliman committed:
fixed the graph area re-rendering 3 times bug

OsamaMSoliman committed:
fixed the pipeline list re-rendering and losing focus when when clicking nodes in the graph area

OsamaMSoliman committed:
adding CustomNode and requirments

le1nux commented:
@OsamaMSoliman is this still work in progress or is this up for review?

cc @vijulshah

OsamaMSoliman committed:
- removed underscores
- every node knows it's parents and children
- fix requirements bug when activePipeline is not set
- more constraints on the nodes, e.g. can't be deleted, can't be connected, etc

OsamaMSoliman committed:
- trying a manual layout, through calc the rows and cols
- handels know the ids of the nodes not just there number
- showing only 1 handle on top but can show more correctly

OsamaMSoliman commented:
@le1nux I just  marked this pull request as ready for review, it still has a problem with overlaying nodes but I'd say let's open a new separate PR for that!

OsamaMSoliman committed:
Merge conflicts of branch 'dev' into introducing_react_flow

vijulshah commented on src/ml_board/frontend/Dashboard/src/components/modelCard/PipelineCard/utils.ts:
```diff
@@ -0,0 +1,63 @@
+
+const g = new Dagre.graphlib.Graph().setDefaultEdgeLabel(() => ({}));
+
+export const getLayoutedNodes = (nodes: Node[], edges: Edge[], options: { direction: string } = { direction: "TB" }) => {
```
I found the layout issue -- why nodes are not laid out properly::

You need to make changes to this function. Here is an example snippet. Please play round to find out the correct positions of x & y.

`

export const getLayoutedNodes = (nodes: Node[], edges: Edge[], options: { direction: string } = { direction: "TB" }) => {
    if (!(nodes.length > 0 && edges.length > 0))
        return { nodes, edges }; // return imidiately if any is emtpy

    g.setGraph({ rankdir: options.direction });
    edges.forEach((edge) => g.setEdge(edge.source, edge.target));
    nodes.forEach((node) => g.setNode(node.id, node as Label));
    Dagre.layout(g);
    let i=0
    return {
        nodes: nodes.map((node: any) => {
            const { x, y } = g.node(node.id);
            // Find nodes with the same rank
            const nodesWithSameRank = nodes.filter((n: any) => n.rank === node.rank);
            // If there are nodes with the same rank (excluding the current node), update x-coordinate
            if (nodesWithSameRank.length > 1) {
                const sameRankNode: any = nodesWithSameRank.find((n) => n.id !== node.id);
                if (sameRankNode) {
                    node.x = sameRankNode.x; 
                    node.position.x = sameRankNode.position.x
                    // ------ PLAY AROUND TO FIND RIGHT CONFIGURATION ------ 
                    i=i+1
                    let node_with_new_corodinates = { ...node, x:x+(100*i), y: y+(100*i), position: { x:x+(100*i), y: y+(100*i) } }
                    return node_with_new_corodinates;
                }
                else {
                    // ------ PLAY AROUND TO FIND RIGHT CONFIGURATION ------ 
                    i=i+1
                    let node_with_new_corodinates = { ...node, x:x+(100*i), y: y+(100*i), position: { x:x+(100*i), y: y+(100*i) } }
                    return node_with_new_corodinates;
                }
            }
            else {
                // ------ PLAY AROUND TO FIND RIGHT CONFIGURATION ------ 
                i=i+1
                let node_with_new_corodinates = { ...node, x:x+(100*i), y: y+(100*i), position: { x:x+(100*i), y: y+(100*i) } }
                return node_with_new_corodinates;
            }
        }),
        edges: edges
    };
};

`

OsamaMSoliman replied:
optimization aside this code is not solving the problem for me, should I change something else somewhere ??
and why `100`? do you assume that a node width and height is 100px?

vijulshah reviewed (changes requested):
I found the nodes layout issue in the graph & I have added the comment above on how to resolve it. It would be better that we solve it & then merge it

vijulshah commented:
I found the nodes layout issue in the graph & I have added the comment above on how to resolve it. It would be better that we solve it & then merge it

OsamaMSoliman commented on src/ml_board/frontend/Dashboard/src/components/modelCard/ModelCard.tsx:
```diff
@@ -196,16 +197,16 @@ export default function ModelCard() {
         link.download = `Experiment_${experiment_id}_ModelCard`;
-        link.click();   
-        
+        link.click();
```
I'm still checking this one out!

OsamaMSoliman commented on src/ml_board/frontend/Dashboard/src/components/modelCard/PipelineCard/index.tsx:
```diff
@@ -0,0 +1,56 @@
+                    <GraphArea />
+                </ReactFlowProvider>
+            </Grid>
+            <Grid item container xs="auto" direction="column">
```
actually I like these ideas 👍🏼 but maybe can we implement then in a new PR?

OsamaMSoliman commented on src/ml_board/frontend/Dashboard/src/components/modelCard/flowGraph/FlowGraph.tsx:
yes I deleted the folder, and I believe some other files are obsolete too, I'll check and delete them too

OsamaMSoliman committed:
deleted FlowGraph folder

OsamaMSoliman committed:
- removed old files
- fixed the saving functionality

Commit:
Pipeline Graph Node positions corrected, created custom node and edges in a simple way

Commit:
Correcting Model Card Pipeline Graph and its Components

Commit:
Css and Variable type changes left

Commit:
Added variable types, corrected css and commented console.logs

Commit:
Added variable types, corrected css and commented console.logs

Commit:
Model Card Css bug fixed
